src/Passes/Source/Rules/OperandPrototype.cpp

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "Rules/OperandPrototype.hpp"
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    IOperandPrototype::~IOperandPrototype() = default;
+    bool IOperandPrototype::matchChildren(Value* value, Captures& captures) const
+    {
+        auto user = llvm::dyn_cast<llvm::User>(value);
+        if (!children_.empty())
+        {
+            if (user == nullptr)
+            {
+                return false;
+            }
+
+            if (user->getNumOperands() != children_.size())
+            {
+                return false;
+            }
+
+            uint64_t i = 0;
+            while (i < children_.size())
+            {
+                auto v = user->getOperand(static_cast<uint32_t>(i));
+                if (!children_[i]->match(v, captures))
+                {
+                    return false;
+                }
+                ++i;
+            }
+
+            return true;
+        }
+
+        // TODO(QAT-private-issue-33): value may be other type than llvm::User. Check other relevant types
+        // and deal with it.
+
+        return true;
+    }
```

Comment:
(I can be wrong, I don't fully understand the logic)
I see that `IOperandPrototype::matchChildren()` calls `match()` for all the children. I expected it to call `match()` for the current node also (in both cases before returning `true`).

Reply:
`match()` calls `matchChildren()` (if required by the pattern), then `matchChildren()` calls `match` on all the children which, in turn, calls `matchChildren()` on their children until the pattern is exhausted. I hope this clarifies things. `match` is responsible for testing whether the current node matches the given operand where as `matchChildren` tests all the children. 

I hope this clarifies the logic. If not feel free to ask again.

Reply:
Makes sense. Consider adding 

a comment to the description of `match()`, something like
> /// Is expected to validate the current node and call `matchChildren()` if needed by the pattern.

and a comment to the description of `matchChildren()`, something like
> /// Is expected to be called by `match()`.

Reply:
I've updated the text in an attempt to make this more clear.